keps/sig-network/20200130-Uncapped-IPv6.md

```diff
@@ -0,0 +1,316 @@
+`--cidr-allocator-type="RangeAllocator"`
+https://github.com/kubernetes/kubernetes/blob/ded6ee953c68f8333ee6291e0bcb7e58604fac00/pkg/controller/nodeipam/ipam/doc.go#L17-L29
+
+The Node IPAM controller manages the assigned CIDR ranges to each node.
+It uses a CidrSet to manage a set of CIDR ranges from which blocks of IPs can be allocated from.
+A new CidrSet struct using roaring bitmaps will be added:
+
+```go
+// CidrSet manages a set of CIDR ranges from which blocks of IPs can
+// be allocated from.
+// It allocates the IP subnet addresses in a bitmap as an offset of the ClusterCIDR IP subnet
+// The masks of the ClusterCIDR and NodeCIDRs are known in advance
+type CidrSet struct {
+  sync.Mutex
+  clusterCIDR   *net.IPNet
+  subnetMask    net.IPMask
+  base          *big.Int
+  offset        *big.Int
+  maxCIDRs      int
+  nextCandidate int
+  used          *roaring.Bitmap
+}
+```
+
+### User Stories
+
+#### Story 1
+
+As a Kubernetes user I want to be able to choose any subnet mask without restrictions for my Nodes or Service IPs.
+
+#### Story 2
+
+As a Kubernetes developer I want to reduce the size of memory used for Service IPs and CIDRs Ranges allocation
+without any significant penalty on performance.
+
+### Implementation Details/Notes/Constraints
+
+PR with a prototype here:
+
+https://github.com/kubernetes/kubernetes/pull/86620
+
+Introduce a new feature gate Roaring?? that allow to use the new bitmap structures.
+
+### Risks and Mitigations
+
+The roaring bitmaps have a limited range of 2^32 integer that seems big enough.
+This implies that user won't be able to use all the ranges configured.
+
+The [range limitation can be removed if needed](https://github.com/RoaringBitmap/CRoaring/issues/1#issuecomment-168721352), but the effort doesn't seem worth the value and the data stored can be too big.
+
+Other option is to relax current limitations to the new limit given by the roaring bitmaps.
+
+## Design Details
+
+### Test Plan
+
+* Benchmark tests comparing performance, memory and CPU usage
+* Scalability tests (TBD)
+* Upgrade tests
+
+### Graduation Criteria
+
+Alpha -->  Beta
+ 
+- TBD
+
+Beta --> GA
+
+- TBD
+
+### Upgrade / Downgrade Strategy
+
+- The ipallocator will need a new prefix in the etcd registry to avoid conflicts with previous
+or older bitmaps versions.
+Once enabled the new bitmap storage, the repair loop will recreate the right entries in the new storage.
+
+- The ipam controller store the status on the node spec fields
+
+### Version Skew Strategy
+
+- The ipallocator will need a new prefix in the etcd registry to avoid conflicts with previous
+or older bitmaps versions.
+Once enabled the new bitmap storage, the repair loop will recreate the right entries in the new storage.
+
+- The ipam controller store the status on the node spec fields
+
+## Drawbacks
+
+Current limits are not very restrictive and the lack of IPv6 deployments doesn't allow to evaluate how really
```

Comment:
This.  I am worried that this feels good, but doesn't solve a real problem.

Reply:
exactly that's why I created the KEP, to check if this is really a problem ... I would not move forward without any use case or evidence that this is going to fix "something" :sweat_smile: